src/classes/ADDR_Addresses_TDTM.cls

```diff
@@ -1088,6 +1076,88 @@ public class ADDR_Addresses_TDTM extends TDTM_Runnable {
+    }
+
+
+    /*******************************************************************************************************
+    * @description Updates the Address record's MailingCountry field when it is empty
+    *              with the Account's BillingCountryCode or Contact's MailingCountryCode value.
+    * @param sobjSrc the source Account or Contact
+    * @param strFieldPrefix the address fields to copy to, ie., Mailing, Other, Billing, Shipping
+    * @param addr the Address object to update accordingly
+    * @return void
+    */
+    @testVisible
+    private static void updateEmptyAddressMailingCountryWithAccountOrContactCountryCode(sObject sobjSrc, String strFieldPrefixSrc, Address__c addr) {
+        // Lookup Country if needed, by Country Code
+        String strCountryCode = (String)sobjSrc.get(strFieldPrefixSrc + 'CountryCode');
+
+        if (String.isNotBlank(strCountryCode) && String.isEmpty(addr.MailingCountry__c)) {
+            addr.MailingCountry__c = validCountriesByCode.get(strCountryCode.toUpperCase());
+        }
+    }
+
+
+    /*******************************************************************************************************
+    * @description Updates the Address record's MailingState fields when it is empty
+    *              and/or mismatched with the Account's BillingStateCode or Contact's MailingStateCode value.
+    * @param sobjSrc the source Account or Contact
+    * @param strFieldPrefix the address fields to copy to, ie., Mailing, Other, Billing, Shipping
+    * @param addr the Address object to update accordingly
+    * @return void
+    */
+    @testVisible
+    private static void updateEmptyOrMismatchedAddressMailingStateWithAccountOrContactStateCode(sObject sobjSrc, String strFieldPrefixSrc, Address__c addr) {
+        // Lookup State if needed, by Country Code and State Code.
+        String strStateCode = (String)sobjSrc.get(strFieldPrefixSrc + 'StateCode');
+
+        if (String.isNotBlank(strStateCode)
+            && (String.isEmpty(addr.MailingState__c) || !isMatchingStateCode(strStateCode, addr.MailingState__c))) {
+
+            // Assume it's a valid statecode
+            addr.MailingState__c = strStateCode;
+        }
+    }
+
+
+    /*******************************************************************************************************
+    * @description Determines if the Contact's or Account's state code matches the Addresses Mailing State
+    *              by looking up
+    * @param strStateCode the Contact's or Account's state code
+    * @param addr the Address object to update accordingly
+    * @return void
+    */
+    @testVisible
+    private static Boolean isMatchingStateCode(String strStateCode, String addressMailingState) {
+        Boolean isMatchingStateCode = false;
+
+        if(String.isNotBlank(addressMailingState)
+            && validStatesByLabel != null && validStatesByLabel.containsKey(addressMailingState.toUpperCase())) {
```

Comment:
- Should we exit fast if the `addressMailingState` is blank?
- What is the cost of checking and getting the value `validStatesByLabel.get(addressMailingState.toUpperCase())` and calling `toUpperCase()` twice?
- Should we use `String.equalsIgnoreCase` instead of `String.equals`?

For example, the code can be written as (*this has not been tested*):

        if (String.isBlank(addressMailingState)) {
            return false;
        }

        String mappedMailingState = validStatesByLabel == null
            ? ''
            : validStatesByLabel.get(addressMailingState.toUpperCase());
        
        return strStateCode.equalsIgnoreCase(mappedMailingState);

Reply:
- Yes, exiting fast if the `addressMailingState` is blank sounds good.
- Agreed, the code can remove the repeated call to `toUpperCase()`. Thanks for calling this out, I just read a post about the `containsKey()` CPU usage:
 (https://salesforce.stackexchange.com/questions/179661/map-of-get-vs-containskey-efficient-check)
- Yes, `String.equalsIgnoreCase` would work here. Thanks!